x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/AbstractGeoHexGridTiler.java

```diff
@@ -0,0 +1,190 @@
+        if (bounds.maxX() - bounds.minX() < 180d) {
+            final long minH3 = H3.geoToH3(bounds.minY(), bounds.minX(), 0);
+            final long maxH3 = H3.geoToH3(bounds.maxY(), bounds.maxX(), 0);
+            if (minH3 == maxH3) {
+                valueIndex = setValuesByRecursion(values, geoValue, minH3, 0, valueIndex);
+                for (long n : H3.hexRing(minH3)) {
+                    valueIndex = setValuesByRecursion(values, geoValue, n, 0, valueIndex);
+                }
+                return valueIndex;
+            }
+            // TODO: specialize when they are neighbour cells.
+        }
+        for (long h3 : RES0CELLS) {
+            valueIndex = setValuesByRecursion(values, geoValue, h3, 0, valueIndex);
+        }
+        return valueIndex;
+    }
+
+    /**
+     * Recursively search the H3 tree, only following branches that intersect the geometry.
+     * Once at the required depth, then all cells that intersect are added to the collection.
+     */
+    private int setValuesByRecursion(
+        GeoShapeCellValues values,
+        GeoShapeValues.GeoShapeValue geoValue,
+        long h3,
+        int precision,
+        int valueIndex
+    ) throws IOException {
+        final GeoRelation relation = relateTile(geoValue, h3);
+        if (precision == this.precision) {
+            // When we're at the desired level
+            return maybeAdd(h3, relation, values, valueIndex);
+        } else {
+            assert precision < this.precision;
+            // When we're at higher tree levels, check if we want to keep iterating.
+            if (relation != GeoRelation.QUERY_DISJOINT) {
+                int i = 0;
+                if (relation == GeoRelation.QUERY_INSIDE) {
+                    // H3 cells do not fully contain the children. The only one we know we fully contain
+                    // is the center child which is always at position 0.
+                    final long centerChild = H3.childPosToH3(h3, i++);
+                    valueIndex = setAllValuesByRecursion(values, centerChild, precision + 1, valueIndex, valueInsideBounds(geoValue));
+                }
+                final int numChildren = H3.h3ToChildrenSize(h3);
+                for (; i < numChildren; i++) {
+                    final long child = H3.childPosToH3(h3, i);
+                    valueIndex = setValuesByRecursion(values, geoValue, child, precision + 1, valueIndex);
+                }
+                // H3 cells do intersects with other cells that are not part of the children cells. If the parent cell of those
+                // cells is disjoint, they will not be visited, therefore visit them here.
+                final int numNoChildren = H3.h3ToNotIntersectingChildrenSize(h3);
+                for (int j = 0; j < numNoChildren; j++) {
+                    final long noChild = H3.noChildIntersectingPosToH3(h3, j);
+                    if (relateTile(geoValue, H3.h3ToParent(noChild)) == GeoRelation.QUERY_DISJOINT) {
```

Comment:
Is a trade off between having to keep things in memory and the possibility of calling relateTile twice. Would be a nice optimisation but it will be hard :)